addons/sale_loyalty/models/sale_order.py

```diff
@@ -989,10 +993,13 @@ def _try_apply_code(self, code):
                 return {'error': _('This coupon has already been used.')}
             program = coupon.program_id
 
+        today = fields.Date.context_today(self)
         if not program or not program.active:
             return {'error': _('This code is invalid (%s).', code), 'not_found': True}
+        elif (program.date_from and program.date_from > today):
+            return {'error': _('This code is not yet active (%s).', code)}
```

Comment:
I just realized, but... Why are these lines needed? You either pass through `_get_trigger_domain` for the coupon or through `_get_program_domain` for the program, both these searches includes the `date_from` and `date_to`.
How can you get to these lines? Aren't they dead code? :thinking:

Reply:
```python
    def _try_apply_code(self, code):
    
        base_domain = self._get_trigger_domain()
        domain = expression.AND([base_domain, [('mode', '=', 'with_code'), ('code', '=', code)]])
        rule = self.env['loyalty.rule'].search(domain)
        program = rule.program_id  # OK already filtered by the domain

        if not program:
            coupon = self.env['loyalty.card'].search([('code', '=', code)])
            program = coupon.program_id  # NOT OK, could be outside validity period
```

Not dead code :stuck_out_tongue:

Reply:
...

```python

        if not program:
            coupon = self.env['loyalty.card'].search([('code', '=', code)])
            if not coupon.program_id.filtered_domain(self._get_program_domain()):
                return {'error': _('This code is invalid (%s).', code), 'not_found': True}
            program = coupon.program_id
```
Dead code? :thinking: (It's that time of the year again, I'm melting)

Reply:
Ok, you win :ok_hand: 
(We're all (mad and) melting here)